Added rio-clip

brendan-ward opened the pull request:
Adds support for ```rio clip``` as per comments in #323

perrygeo commented:
The issue with negative xmin being interpreted as an option is something I hadn't thought of but I think adding a space is a reasonable work around.

It's looking good, I get the exact same output as `gdal_translate -projwin`.
On one hand that's good but it can be slightly problematic in that the bbox of the resulting raster does not necessarily fully "cover" the specified bbox. 

```
>>> src.window_bounds(src.window(*bbox))
(9.474999969248074, 46.44166664920006, 16.974999968048053, 49.04166664878406)
>> bbox
[9.479969516649021, 46.43181732846955, 16.979666782304037, 49.03907420510758]
>>> import shapely
>>> calcbox = shapely.geometry.box(*src.window_bounds(src.window(*bbox)))
>>> origbox = shapely.geometry.box(*bbox)
>>> calcbox.covers(origbox)
False
```
<img width="539" alt="screen shot 2015-08-21 at 9 15 32 am" src="https://cloud.githubusercontent.com/assets/1151287/9409178/7498c298-47e6-11e5-9ed4-be06e4251e1b.png">

This can be problematic in certain cases: grab the bbox of a point feature collection, clip the raster to that bbox, query raster values at each point and some of points along the edge may not intersect with the raster surface.

The current behavior probably works for most cases though and changing it would require rethinking `src.window` to include a `fully_cover` keyword argument or something (effectively an "all-touched" option for creating windows)

geowurster commented:
@brendan-ward I have not caught up on #323 so forgive me if this was covered, but another method for fetching and parsing the bbox is a callback paired with a `click.option(default='-')`, which I think has more flexibility moving forward and removes the need for a leading space.

```python
import sys
import json

def _cb_bounds(ctx, param, value):
    if value == '-':
        v = sys.stdin.read().strip()
    else:
        v = value
    
    # Everything returns a 4 element tuple so no parsing has to be done in the command function.
    if v.startswith('['):
        return json.loads(v)
    else:
        # regex filter, cast to int
   
@click.option('--bounds', default='-', callback=_cb_bounds)
```

brendan-ward commented:
@perrygeo point taken; I went for consistency w/ existing behavior in rasterio.  Should this be a parameter for rio-clip?  Or should we instead make sure that the window we use in this case always fully covers the input bounds?   Either way - when the bounds don't match the pixel boundaries, the output will either be slightly larger or slightly smaller than requested (probably a good thing to note in docs & docstring; I'll do so).

@geowurster good suggestion - I was following @sgillies [comment](https://github.com/mapbox/rasterio/pull/323#issuecomment-133158182) on #323 but agreed that it introduces unexpected awkwardness around leading space ahead of negative xmins.  I'm all for a callback approach - do you know if we can make it work for ```click.argument``` as well as ```click.option```?  I only saw the [option examples](http://click.pocoo.org/4/advanced/#parameter-modifications)...

perrygeo commented:
@brendan-ward I think exposing the parameter in rio-clip would be great but the functionality could go in [`DatasetReader.window`](https://github.com/mapbox/rasterio/blob/2120a07111998dd18f0911c9cede4be6fe09797a/rasterio/_base.pyx#L385) to allow other methods to use this strategy for generating windows that are guaranteed to cover. I don't think it should block this PR, though - I'll make a separate issue.

geowurster commented:
@brendan-ward `click.argument()` can have a callback as well, but the leading space is still an issue because [callbacks happen after argument parsing](http://click.pocoo.org/5/advanced/#callback-evaluation-order).

I think there are some UI and "down the road" reasons to use an option for this as well, unless this command is only ever intended to clip to a bbox, in which case maybe `rio crop` is a better name?  But even then, the input/output files are (I think) the only positional arguments across all `rio` commands.

The `--bounds` option in `rio.options` requires 4 positional arguments, but the same flag for this command will behave differently.  I'm wondering if maybe we don't want a `--bounds` and `--json-bounds` or something to prevent a situation where the usage for the same flag is different on multiple commands even though the intended effect is the same.

brendan-ward commented:
@geowurster If we used a parameter here, we'd likely want to call it ```--bbox``` instead of bounds.  Agreed that multiple variants of ```--bounds``` is problematic.  See [comments](https://github.com/mapbox/rasterio/pull/323#issuecomment-133144431) in #323 onward...

My sense is the intended usage here still justifies bbox as an argument rather than option, if possible.  What are the UI / down the road issues around that approach?

geowurster commented:
@brendan-ward et al. - let me preface this by saying I feel like I may be bike shedding so in the end I'll defer to consensus.

I thought about this for a bit, looked over #323, get where everyone is coming from, and can't come up with any concrete "down the road" problems since we have `rio mask`, so I think what's really bothering me is the leading space, quotes and the slight UI difference even though the intended functionality is the same.

What about further relying on more unix commands?  We could have `--bounds` default to `- - - -` for reading from `stdin` and relying on `$()` commands for transforming our various input formats to 4 values separate values.  If we handle the validation and transformation in a callback it doesn't matter what we give it, as long as it receives 4 values.  By effectively doing `"[w, s, e, n]".split()` and `"w, s, e, n".split()` with unix commands we can quietly handle the transformation:

```python
def _cb_bounds(ctx, param, value):

    if value == '- - - -':
        value = sys.stdin.read().split()
    
    # This can probably be simplified
    bounds = ()
    for v in value:
        for c in ('[', ']', ','):
            v = v.strip(c)
        bounds.append(float(v))
    
    return bounds
```

So now we require the user to do a little more work on the command line to support the two additional bounds syntaxes:

```console
$ echo "w, s, e, n" | rio clip infile.tif outfile.tif
$ something | jq ".bounds" | rio clip infile.tif outfile.tif  # receives [w, s, e, n] from stdin
$ rio clip infile.tif outfile.tif --bounds $(fio info --bounds map.geojson)
$ rio clip infile.tif outfile.tif --bounds $(echo "[w, s, e, n]")
$ rio clip infile.tif outfile.tif --bounds $(echo $(jq ".bounds" file.json))
```

It's not pretty but it is unixy.

sgillies commented:
I think getting the bounds from stdin is off track.

For commands that are filters, stdin is important to support. rio-clip isn't a filter. Besides, the primary argument is a raster file and we can't get those from stdin (except maybe GeoTIFFs in GDAL 2.0, but that's another topic). The bounds input is important but it's not the primary, the command could work without it (being an boutique `cp`).

brendan-ward commented:
@sgillies so I'm sort of stuck at this point.  I'm not interested in rio-clip being a boutique ```cp```; if there are no bounds or bounds are outside of the extent of the input, I want this to fail with an error.

Is the issue with taking from stdin more around it as an argument rather than an option, or around defaulting to stdin as suggested by @perrygeo in #323?  We could instead leave this with ```default=None``` and expect user to provide it or explicitly use the ```-``` placeholder to read from stdin (which seems to be required no matter how we use it, argument or option) but prefer it be entered directly.

Seems worth a look again at how we can leverage [click](http://click.pocoo.org/5/advanced/#token-normalization) to handle variable parameters.  Ideally, use of ```--bounds``` or parameter of similar intent be handled the same way throughout rio.

sgillies commented:
@brendan-ward sorry to block you!

I'm fine with bounds as an argument. And fine with it being required. We don't need to get bounds from stdin since rio-clip isn't a filter.

Related: I typed up the current state of CRS handling at https://github.com/mapbox/rasterio/wiki/CLI-design.

brendan-ward commented:
Updated to use ```--bounds``` and drop use of stdin.  Added back in ```--like``` because I suspect we won't regret it later, even if there are multiple ways to realize the same result.

This better?

```
Usage: rio clip [OPTIONS] INPUT OUTPUT

  Clips a raster using bounds (in same crs as input raster) or the bounds of
  a template raster (bounds will be transformed to match the input).

    $ rio clip input.tif output.tif --bounds xmin ymin xmax ymax
    $ rio clip input.tif output.tif --like template.tif

  It can also be combined to read bounds of a feature dataset using Fiona:

    $ rio clip input.tif output.tif --bounds $(fio info features.shp --bounds)

Options:
  -o, --output PATH            Path to output file (optional alternative to a
                               positional arg for some commands).
  --bounds FLOAT...            Output bounds: left bottom right top.
  --like PATH                  Raster dataset to use as a template for bounds
  -f, --format, --driver TEXT  Output format driver
  --co NAME=VALUE              Driver specific creation options.See the
                               documentation for the selected output driver
                               for more information.
  --help                       Show this message and exit.
```


@perrygeo does something like this get you close enough to your original need:

```$ rio clip input.tif output.tif --bounds $(fio info features.shp --bounds)```

Many thanks to @geowurster for that expression!

perrygeo commented:
@brendan-ward  That sounds good to me - we can get bounds in programmatically and it's consistent with the rest of the bounds handling.

sgillies commented:
:+1: this complements rio-convert very well.

brendan-ward commented:
Ok - this is ready for review in its current form.

sgillies commented on docs/cli.rst:
```diff
@@ -19,6 +19,7 @@ Rasterio's command line interface is a program named "rio".
     Commands:
       bounds     Write bounding boxes to stdout as GeoJSON.
       calc       Raster data calculator.
+      clip       Clip a raster using a bounding box.
```
How about "Clip a raster to given bounds" since there's a `--bounds` option?

sgillies commented on docs/cli.rst:
```diff
@@ -157,6 +158,27 @@ efficiently in Python.
+Added in 0.27
+
+The ``clip`` command clips a raster using bounds (in same crs as input raster)
+or the bounds of a template raster (transformed to match the crs of the input).
```
"clips a raster to given bounds... or the bounds of a template raster..."

sgillies commented on rasterio/rio/features.py:
```diff
@@ -17,10 +18,96 @@
+
+            with rasterio.open(output, 'w', **out_kwargs) as out:
+                for bidx in range(1, src.count + 1):
+                    out.write_band(bidx, src.read(bidx, window=window))
```
Should be able to do just one read and one write:

```
with rasterio.open(output, 'w', **out_kwargs) as dst:
    dst.write(src.read(window=window))
```

sgillies commented on setup.py:
```diff
@@ -235,6 +235,7 @@ def copy_data_tree(datadir, destdir):
         [rasterio.rio_commands]
         bounds=rasterio.rio.features:bounds
         calc=rasterio.rio.calc:calc
+        clip=rasterio.rio.features:clip
```
I feel like clip could go in rasterio.rio.convert since rio-convert is so closely related: they are two commands that are joined in gdal_translate.

brendan-ward commented:
Refactored per feedback, and squashed commits...  

Please take a look at ```disjoint_bounds``` in ```coords.py``` - I wasn't sure where to put this, originally it was only used in ```rio/features.py``` but is now needed for rio-clip in ```rio/convert.py``` too.

sgillies commented:
Excellent.
